File: src/plugins/executor.rs
use std::future::Future;
use std::path::PathBuf;
use std::pin::Pin;
use std::process::Stdio;

use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;

use crate::plugins::registry::Plugin;

/// Result from a CLI execution: the response text and optional session ID for resume.
pub struct CliResult {
    pub text: String,
    pub session_id: Option<String>,
}

// ── CLI-based executor ───────────────────────────────────────────────────────

/// Execute a plugin by spawning the `claude` CLI as a subprocess.
///
/// The CLI has full access to: Bash, Read, Grep, Glob, Agent, MCP servers
/// (Coralogix, Trino, Splitz, etc.), and all native Claude Code tools.
///
/// This replaces the old Vertex AI API call + 4 custom tools approach.
pub async fn execute_plugin_via_cli(
    plugin: &Plugin,
    query: &str,
    session_id: Option<&str>,
    mcp_config_path: Option<&str>,
) -> Result<CliResult, String> {
    let mut cmd = Command::new("claude");

    // Core flags — use stream-json to capture actual response text
    cmd.arg("--print")
        .arg("--dangerously-skip-permissions")
        .arg("--output-format").arg("stream-json")
        .arg("--verbose")
;

    // System prompt = the plugin's SKILL.md content
    cmd.arg("--system-prompt").arg(&plugin.system_prompt);

    // MCP server configuration
    if let Some(mcp_path) = mcp_config_path {
        let path = PathBuf::from(mcp_path);
        if path.exists() {
            cmd.arg("--mcp-config").arg(mcp_path);
        }
    }

    // Resume existing session for multi-turn conversations
    if let Some(sid) = session_id {
        cmd.arg("--resume").arg(sid);
    }

    // Pass query via stdin to avoid shell escaping issues
    // (enriched queries contain ---, DevRev content, URLs, etc.)
    cmd.arg("--print");

    cmd.stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    tracing::info!(
        fqn = %plugin.fqn,
        query_len = query.len(),
        has_session = session_id.is_some(),
        "Spawning claude CLI for plugin"
    );

    let mut child = cmd.spawn().map_err(|e| {
        format!("Failed to spawn claude CLI: {}. Is claude installed and in PATH?", e)
    })?;

    // Write query to stdin
    if let Some(mut stdin) = child.stdin.take() {
        use tokio::io::AsyncWriteExt;
        stdin.write_all(query.as_bytes()).await.map_err(|e| {
            format!("Failed to write query to claude CLI stdin: {}", e)
        })?;
        drop(stdin); // Close stdin to signal EOF
    }

    let output = child.wait_with_output().await.map_err(|e| {
        format!("Claude CLI process failed: {}", e)
    })?;

    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);

    if !output.status.success() && stdout.is_empty() {
        tracing::error!(
            fqn = %plugin.fqn,
            exit_code = ?output.status.code(),
            stderr = %stderr,
            "Claude CLI exited with error"
        );
        return Err(format!("Claude CLI error (exit {}): {}",
            output.status.code().unwrap_or(-1), stderr));
    }

    // Parse stream-json output — each line is a JSON object
    // Collect text from "assistant" messages and metadata from "result" message
    let mut text = String::new();
    let mut cli_session_id: Option<String> = None;
    let mut cost: f64 = 0.0;
    let mut duration_ms: u64 = 0;

    for line in stdout.lines() {
        if line.is_empty() { continue; }
        let parsed: serde_json::Value = match serde_json::from_str(line) {
            Ok(v) => v,
            Err(_) => continue,
        };

        let msg_type = parsed.get("type").and_then(|v| v.as_str()).unwrap_or("");

        match msg_type {
            "assistant" => {
                // Extract text from assistant message content blocks
                if let Some(content) = parsed.pointer("/message/content").and_then(|c| c.as_array()) {
                    for block in content {
                        if let Some(t) = block.get("text").and_then(|t| t.as_str()) {
                            text.push_str(t);
                        }
                    }
                }
            }
            "result" => {
                // Capture metadata from the final result line
                if let Some(r) = parsed.get("result").and_then(|v| v.as_str()) {
                    if !r.is_empty() && text.is_empty() {
                        text = r.to_string();
                    }
                }
                cli_session_id = parsed.get("session_id")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string());
                cost = parsed.get("total_cost_usd")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.0);
                duration_ms = parsed.get("duration_ms")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0);
            }
            _ => {}
        }
    }

    tracing::info!(
        fqn = %plugin.fqn,
        text_len = text.len(),
        session_id = ?cli_session_id,
        cost_usd = cost,
        duration_ms = duration_ms,
        "Claude CLI execution complete"
    );

    Ok(CliResult {
        text,
        session_id: cli_session_id,
    })
}

/// Execute a plugin by spawning the `claude` CLI with streaming JSON output.
///
/// Streams text chunks as they arrive, calling `on_text` for each chunk.
/// Returns the final result text and session ID.
pub async fn execute_plugin_via_cli_streaming<F>(
    plugin: &Plugin,
    query: &str,
    session_id: Option<&str>,
    mcp_config_path: Option<&str>,
    mut on_text: F,
) -> Result<CliResult, String>
where
    F: FnMut(&str) + Send,
{
    let mut cmd = Command::new("claude");

    // Core flags — use stream-json for streaming
    cmd.arg("--print")
        .arg("--dangerously-skip-permissions")
        .arg("--output-format").arg("stream-json")
        .arg("--verbose")
;

    // System prompt = the plugin's SKILL.md content
    cmd.arg("--system-prompt").arg(&plugin.system_prompt);

    // MCP server configuration
    if let Some(mcp_path) = mcp_config_path {
        let path = PathBuf::from(mcp_path);
        if path.exists() {
            cmd.arg("--mcp-config").arg(mcp_path);
        }
    }

    // Resume existing session
    if let Some(sid) = session_id {
        cmd.arg("--resume").arg(sid);
    }

    // Pass query via stdin (avoid shell escaping issues with enriched content)
    cmd.stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    tracing::info!(
        fqn = %plugin.fqn,
        query_len = query.len(),
        has_session = session_id.is_some(),
        "Spawning claude CLI (streaming) for plugin"
    );

    let mut child = cmd.spawn().map_err(|e| {
        format!("Failed to spawn claude CLI: {}. Is claude installed and in PATH?", e)
    })?;

    // Write query to stdin
    if let Some(mut stdin) = child.stdin.take() {
        use tokio::io::AsyncWriteExt;
        stdin.write_all(query.as_bytes()).await.map_err(|e| {
            format!("Failed to write query to claude CLI stdin: {}", e)
        })?;
        drop(stdin); // Close stdin to signal EOF
    }

    let stdout = child.stdout.take()
        .ok_or("Failed to capture claude CLI stdout")?;

    let mut reader = BufReader::new(stdout).lines();
    let mut full_text = String::new();
    let mut cli_session_id: Option<String> = None;

    // Read streaming JSON lines
    while let Some(line) = reader.next_line().await.map_err(|e| {
        format!("Error reading claude CLI output: {}", e)
    })? {
        if line.is_empty() {
            continue;
        }

        // Parse each JSON line
        let parsed: serde_json::Value = match serde_json::from_str(&line) {
            Ok(v) => v,
            Err(_) => continue, // Skip unparseable lines
        };

        let msg_type = parsed.get("type").and_then(|v| v.as_str()).unwrap_or("");

        match msg_type {
            "assistant" => {
                // Extract text from assistant message content blocks
                if let Some(message) = parsed.get("message") {
                    if let Some(content) = message.get("content").and_then(|c| c.as_array()) {
                        for block in content {
                            if let Some(text) = block.get("text").and_then(|t| t.as_str()) {
                                full_text.push_str(text);
                                on_text(text);
                            }
                        }
                    }
                }
            }
            "result" => {
                // Final result — capture session ID and result text
                if let Some(result_text) = parsed.get("result").and_then(|v| v.as_str()) {
                    if !result_text.is_empty() {
                        full_text = result_text.to_string();
                    }
                }
                cli_session_id = parsed.get("session_id")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string());

                let cost = parsed.get("total_cost_usd")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.0);

                tracing::info!(
                    fqn = %plugin.fqn,
                    text_len = full_text.len(),
                    session_id = ?cli_session_id,
                    cost_usd = cost,
                    "Claude CLI streaming complete"
                );
            }
            "system" => {
                // Capture session ID from init message
                let subtype = parsed.get("subtype").and_then(|v| v.as_str()).unwrap_or("");
                if subtype == "init" {
                    cli_session_id = parsed.get("session_id")
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());
                }
            }
            _ => {
                // Ignore other message types (tool_use, system hooks, etc.)
            }
        }
    }

    // Wait for the process to finish
    let status = child.wait().await.map_err(|e| {
        format!("Failed to wait for claude CLI: {}", e)
    })?;

    if !status.success() {
        tracing::warn!(
            fqn = %plugin.fqn,
            exit_code = ?status.code(),
            "Claude CLI exited with non-zero status (may still have output)"
        );
    }

    Ok(CliResult {
        text: full_text,
        session_id: cli_session_id,
    })
}

// ── Backward-compatible wrapper ──────────────────────────────────────────────

/// Execute a plugin — backward-compatible wrapper that matches the old signature.
///
/// This is called from events.rs. It delegates to `execute_plugin_via_cli`.
/// The `client` and `registry` params are kept for API compatibility but unused.
pub fn execute_plugin<'a>(
    plugin: &'a Plugin,
    query: &'a str,
    _history: &'a [crate::claude::types::Message],
    _client: &'a crate::claude::client::ClaudeClient,
    _registry: &'a crate::plugins::registry::PluginRegistry,
    _depth: u32,
) -> Pin<Box<dyn Future<Output = Result<String, String>> + Send + 'a>> {
    Box::pin(async move {
        // Resolve MCP config path relative to the executable
        let mcp_config = std::env::current_dir()
            .ok()
            .map(|d| d.join("mcp-servers.json"))
            .and_then(|p| if p.exists() { Some(p.to_string_lossy().to_string()) } else { None });

        let result = execute_plugin_via_cli(
            plugin,
            query,
            None, // No session resume in the backward-compatible path
            mcp_config.as_deref(),
        )
        .await?;

        Ok(result.text)
    })
}

// ── Unit tests ────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cli_result_struct() {
        let result = CliResult {
            text: "Hello".to_string(),
            session_id: Some("abc-123".to_string()),
        };
        assert_eq!(result.text, "Hello");
        assert_eq!(result.session_id.unwrap(), "abc-123");
    }

    #[test]
    fn test_cli_result_no_session() {
        let result = CliResult {
            text: "Response".to_string(),
            session_id: None,
        };
        assert!(result.session_id.is_none());
    }
}

File: src/oac_slack_bot/plugins/executor.py
"""Plugin execution via claude CLI subprocess."""

from __future__ import annotations

import asyncio
import json
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

import structlog

from oac_slack_bot.plugins.registry import Plugin

logger = structlog.get_logger()


@dataclass
class CliResult:
    text: str
    session_id: str | None = None


async def execute_plugin_via_cli(
    plugin: Plugin,
    query: str,
    session_id: str | None = None,
    mcp_config_path: str | None = None,
) -> CliResult:
    """Execute a plugin by spawning the claude CLI subprocess."""
    args = [
        "claude",
        "--print",
        "--dangerously-skip-permissions",
        "--output-format", "stream-json",
        "--verbose",
        "--model", "sonnet",
        "--system-prompt", plugin.system_prompt,
    ]

    if mcp_config_path and Path(mcp_config_path).exists():
        args.extend(["--mcp-config", mcp_config_path])

    if session_id:
        args.extend(["--resume", session_id])

    # Extra --print for stdin mode
    args.append("--print")

    logger.info(
        "spawning_claude_cli",
        fqn=plugin.fqn,
        query_len=len(query),
        has_session=session_id is not None,
    )

    proc = await asyncio.create_subprocess_exec(
        *args,
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    stdout_bytes, stderr_bytes = await proc.communicate(input=query.encode())
    stdout = stdout_bytes.decode(errors="replace")
    stderr = stderr_bytes.decode(errors="replace")

    if proc.returncode != 0 and not stdout:
        logger.error(
            "claude_cli_error",
            fqn=plugin.fqn,
            exit_code=proc.returncode,
            stderr=stderr[:500],
        )
        raise RuntimeError(
            f"Claude CLI error (exit {proc.returncode}): {stderr[:500]}"
        )

    # Parse stream-json output
    text = ""
    cli_session_id: str | None = None
    cost: float = 0.0
    duration_ms: int = 0

    for line in stdout.splitlines():
        if not line.strip():
            continue
        try:
            parsed = json.loads(line)
        except json.JSONDecodeError:
            continue

        msg_type = parsed.get("type", "")

        if msg_type == "assistant":
            content = (parsed.get("message") or {}).get("content", [])
            if isinstance(content, list):
                for block in content:
                    if isinstance(block, dict) and "text" in block:
                        text += block["text"]

        elif msg_type == "result":
            result_text = parsed.get("result", "")
            if result_text and not text:
                text = result_text
            cli_session_id = parsed.get("session_id")
            cost = parsed.get("total_cost_usd", 0.0)
            duration_ms = parsed.get("duration_ms", 0)

    logger.info(
        "claude_cli_complete",
        fqn=plugin.fqn,
        text_len=len(text),
        session_id=cli_session_id,
        cost_usd=cost,
        duration_ms=duration_ms,
    )

    return CliResult(text=text, session_id=cli_session_id)


async def execute_plugin_via_cli_streaming(
    plugin: Plugin,
    query: str,
    session_id: str | None = None,
    mcp_config_path: str | None = None,
    on_text: Callable[[str], None] | None = None,
) -> CliResult:
    """Execute a plugin with streaming output, calling on_text for each chunk."""
    args = [
        "claude",
        "--print",
        "--dangerously-skip-permissions",
        "--output-format", "stream-json",
        "--verbose",
        "--model", "sonnet",
        "--system-prompt", plugin.system_prompt,
    ]

    if mcp_config_path and Path(mcp_config_path).exists():
        args.extend(["--mcp-config", mcp_config_path])

    if session_id:
        args.extend(["--resume", session_id])

    logger.info(
        "spawning_claude_cli_streaming",
        fqn=plugin.fqn,
        query_len=len(query),
        has_session=session_id is not None,
    )

    proc = await asyncio.create_subprocess_exec(
        *args,
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    # Write query to stdin and close
    if proc.stdin:
        proc.stdin.write(query.encode())
        await proc.stdin.drain()
        proc.stdin.close()

    full_text = ""
    cli_session_id: str | None = None

    # Read stdout line by line
    if proc.stdout:
        while True:
            line_bytes = await proc.stdout.readline()
            if not line_bytes:
                break
            line = line_bytes.decode(errors="replace").strip()
            if not line:
                continue

            try:
                parsed = json.loads(line)
            except json.JSONDecodeError:
                continue

            msg_type = parsed.get("type", "")

            if msg_type == "assistant":
                message = parsed.get("message", {})
                content = message.get("content", [])
                if isinstance(content, list):
                    for block in content:
                        if isinstance(block, dict) and "text" in block:
                            chunk = block["text"]
                            full_text += chunk
                            if on_text:
                                on_text(chunk)

            elif msg_type == "result":
                result_text = parsed.get("result", "")
                if result_text:
                    full_text = result_text
                cli_session_id = parsed.get("session_id")
                cost = parsed.get("total_cost_usd", 0.0)
                logger.info(
                    "claude_cli_streaming_complete",
                    fqn=plugin.fqn,
                    text_len=len(full_text),
                    session_id=cli_session_id,
                    cost_usd=cost,
                )

            elif msg_type == "system":
                subtype = parsed.get("subtype", "")
                if subtype == "init":
                    cli_session_id = parsed.get("session_id")

    await proc.wait()

    if proc.returncode and proc.returncode != 0:
        logger.warning(
            "claude_cli_nonzero_exit",
            fqn=plugin.fqn,
            exit_code=proc.returncode,
        )

    return CliResult(text=full_text, session_id=cli_session_id)


async def execute_plugin(
    plugin: Plugin,
    query: str,
) -> str:
    """Backward-compatible wrapper for execute_plugin_via_cli."""
    mcp_config = None
    try:
        candidate = Path.cwd() / "mcp-servers.json"
        if candidate.exists():
            mcp_config = str(candidate)
    except OSError:
        pass

    result = await execute_plugin_via_cli(
        plugin, query, session_id=None, mcp_config_path=mcp_config
    )
    return result.text
